the role junit2json

mvk opened the pull request:
##### SUMMARY

Add a role that converts junit reports into json.


###### Motivation:

As a part of group's desire to be able to make decisions based on data, we're investing into CI/Test observability. 
The observability effort requires data collection and analysis, currently using Splunk with native data format as JSON + some data tagging/enrichment needs.
Current test reports are junit XML, which therefore needs to be converted and sent to a splunk index.

This role `junit2json` is handling reports data conversion, using the previously merged `json2obj` filter.

After merging it we will submit another one for sending the test reports, it should call this one to convert the data and send the resulting `.json` files as splunk data frames.

##### ISSUE TYPE

- New Feature

##### Tests

1. setup standard `collections/ansible_collections/redhatci/ocp` hierarchy
2. run the playbook: `ansible-playbook collections/ansible_collections/redhatci/ocp/roles/junit2json/tests/test.yml`

---

Test-hints: no-check

shaior commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,100 @@
+  ansible.builtin.set_fact:
+    junit2json_xml_report_hash_path: "{{ xml_report }}.{{ junit2json_hash }}"
+    junit2json_xml_report_hash_curr: "{{ junit2json_xml_report_content | ansible.builtin.hash(junit2json_hash) }}"
+    junit2json_xml_report_hash_old: ""
```
Do we need it here if it's empty?

thekad replied:
why is this conversation resolved? I don't see an answer

shaior commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,100 @@
+  when:
+    - not(junit2json_do_merge)
+
+- name: "Setup JSON report file name [`xml_report.endswith('.xml'): true`]"
```
Can we use this 

```
- name: "Setup JSON report file name
  ansible.builtin.set_fact:
    junit2json_json_report_path: "{{ xml_report | basename | regex_replace('\\.xml$', '.json') if xml_report.endswith('.xml') else xml_report | basename + '.json' }}"
  when:
    - not(junit2json_do_merge)`
```
    
instead of the two "Setup ... " tasks?

shaior commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,100 @@
+  when:
+    - junit2json_xml_report_hash_curr != junit2json_xml_report_hash_old
+
+- name: Write json data to destination
```
For the following two Write tasks, you can use a block with a single when conditional, as they share the same logic and content approx.

shaior replied:
Please reply before resolving.

shaior commented on roles/junit2json/tasks/main.yml:
```diff
@@ -0,0 +1,44 @@
+---
+# tasks file main.yml for role redhatci.ocp.junit2json
+
+- name: Validate some variables
```
I think this is more readable: 

```
- name: Validate required variables
  ansible.builtin.assert:
    that:
      - global_reports_path_patterns | length > 0
      - junit2json_input_reports_list | length > 0
      - junit2json_output_dir | length > 0
    fail_msg: "One or more required variables are missing or empty"
```

shaior replied:
If you agree with what I wrote, please change it.

shaior commented on roles/junit2json/tasks/main.yml:
```diff
@@ -0,0 +1,44 @@
+  loop_control:
+    loop_var: path_item
+
+- name: "Merge JUnit XML reports into single file [`junit2json_do_merge: true`]"
```
Please remove the double quotes as other tasks names.

shaior commented on roles/junit2json/tasks/expand.yml:
```diff
@@ -0,0 +1,47 @@
+  when:
+    - not(_junit2json_path_item_stat.stat.isdir | default(false))
+
+- name: "Find JUnit XML reports under path_item [`stat.isdir: true`]"
```
Please remove the double quotes as other tasks names.

shaior commented on roles/junit2json/tasks/expand.yml:
```diff
@@ -0,0 +1,47 @@
+    path: "{{ path_item }}"
+  register: _junit2json_path_item_stat
+
+- name: "Add path_item as a JUnit XML report [`stat.isdir: false`]"
```
Please remove the double quotes as other tasks names.

shaior commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,100 @@
+  when:
+    - junit2json_do_merge
+
+- name: "Update junit2json_output_report_path [`junit2json_do_merge: false`]"
```
Please remove the double quotes as other tasks names.

shaior commented on roles/junit2json/tasks/merge.yml:
```diff
@@ -0,0 +1,40 @@
+  loop: "{{ _junit2json_found_reports.files | map(attribute='path') }}"
+
+# debug only:
+- name: Print XML reports' filenames data
```
Print list of XML report file names

shaior commented on roles/junit2json/tasks/merge.yml:
```diff
@@ -0,0 +1,40 @@
+- name: Merge multiple JUnit XML files into single consolidated report
+  ansible.builtin.shell:
+    cmd: |
+      junitparser merge "{{ junit2json_reports_list | join('" "') }}" -
```
I think you should check that junitparser is installed first using pip module.

thekad replied:
I agree there should be validations, but I think those validations (specially for things like a required library) should be done at the beginning of the role execution, so it fails fast if requirements aren't met

ramperher commented on roles/junit2json/tasks/main.yml:
```diff
@@ -0,0 +1,44 @@
+  ansible.builtin.assert:
+    that: "{{ condition }}"
+  loop:
+    - global_reports_path_patterns | length > 0
```
This variable should follow the naming convention of the role (be prefixed by `junit2json_` keyword)

ramperher commented on roles/junit2json/tasks/expand.yml:
```diff
@@ -0,0 +1,47 @@
+- name: Setup helper variable
+  ansible.builtin.set_fact:
+    junit2json_find_patterns: "{{ global_reports_path_patterns | map('regex_replace', '^', '*.') | list }}"
+    cacheable: false
```
Var to be prefixed by `junit2json` keyword

ramperher commented on roles/junit2json/tasks/main.yml:
```diff
@@ -0,0 +1,44 @@
+    file: expand.yml
+  loop: "{{ junit2json_input_reports_list }}"
+  loop_control:
+    loop_var: path_item
```
maybe this var could be renamed to `_junit2json_path_item`

ramperher commented on roles/junit2json/tasks/main.yml:
```diff
@@ -0,0 +1,44 @@
+    file: convert.yml
+  loop: "{{ junit2json_reports_list }}"
+  loop_control:
+    loop_var: xml_report
```
maybe to be renamed by `_junit2json_xml_report`

ramperher commented on roles/junit2json/tasks/merge.yml:
```diff
@@ -0,0 +1,40 @@
+    - junit2json_debug
+
+- name: Merge multiple JUnit XML files into single consolidated report
+  ansible.builtin.shell:
```
If just using one single command, please use `command` module instead of `shell`

ramperher commented on roles/junit2json/tests/inventory:
```diff
@@ -0,0 +1 @@
+localhost ansible_connection=local ansible_python_interpreter="/usr/bin/env python"
```
missed a line break at the end of the file

ramperher commented on roles/junit2json/tests/test_junit2obj_simple_input.json:
```diff
@@ -0,0 +1 @@
+"{\n    \"time\": 0.000426228,\n    \"tests\": 3,\n    \"failures\": 0,\n    \"errors\": 0,\n    \"skipped\": 2,\n    \"test_suites\": [\n        {\n            \"name\": \"Performance Addon Operator Reboot\",\n            \"time\": 0.0,\n            \"timestamp\": \"2024-12-12T20:12:23\",\n            \"tests\": 3,\n            \"failures\": 0,\n            \"errors\": 0,\n            \"skipped\ ...
```
missed line break at the end of the file

ramperher commented on roles/junit2json/vars/main.yml:
if this file is commented, could it be removed or could you move this content to the README?

ramperher commented on roles/junit2json/tests/test.yml:
```diff
@@ -0,0 +1,33 @@
+        name: redhatci.ocp.junit2json
+      vars:
+        junit2json_input_reports_list:
+          - "{{ role_path }}/../../tests/unit/data/test_junit2obj_simple_input.xml"
```
- is `role_path` var a requirement for this test?
- where is this file located? this `tests/unit/data...` is not present in this change

mvk replied:
`role_path` is a special built-in variable, I don't need to define it.
The data `tests/unit/data` is under the collection root folder.
(2 folders up `role_path`)

ramperher commented on roles/junit2json/tests/test.yml:
```diff
@@ -0,0 +1,33 @@
+        actual: "{{ lookup('file', playbook_dir + '/test_junit2obj_simple_input.json') | from_json }}"
+    - name: Load expected result to variable expected
+      ansible.builtin.set_fact:
+        expected: "{{ lookup('file', playbook_dir + '/../../../tests/unit/data/test_junit2obj_simple_result.json') }}"
```
same regarding the file path

ramperher replied:
got it, thanks, nothing to be changed here

ramperher commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,100 @@
+
+- name: Update global_json_reports_list
+  ansible.builtin.set_fact:
+    global_json_reports_list: "{{ global_json_reports_list | default([]) + [junit2json_output_report_path] }}"
```
you should prefix this var with `junit2json` if possible

mvk replied:
only role's variables should have role specific prefix (`junit2json_` here)
`global_` is a legal prefix for cross-role variables.

ramperher replied:
Got it, thanks.
Taking that into account, where is this global variable intended to be used? Is it going to be consumed by another role from the collections?

thekad replied:
I also don't see an answer to this question, please answer before resolving the convos.
More feedback: you could put `global_json_reports_list` in your defaults file and you won't have to use `default()` here or pass `[]` as a variable

thekad commented on roles/junit2json/README.md:
```diff
@@ -0,0 +1,308 @@
+  
+  
+
+  - **junit2json_input_merged_report**
```
this should have a sane default

thekad replied:
this still doesn't have a sane default

thekad commented on roles/junit2json/README.md:
```diff
@@ -0,0 +1,308 @@
+  
+  
+
+  - **junit2json_output_merged_report**
```
this should also have a sane default

thekad commented on roles/junit2json/README.md:
```diff
@@ -0,0 +1,308 @@
+- name: "Example playbook to use the role redhatci.ocp.junit2json role"
+  hosts: localhost
+  vars:
+    global_json_reports_list: []
```
this vars line is a bad example, because the nature of passing vars to a role makes it so every task inside the role has the same var definition block, every task receives an empty list for this variable, please remove

mvk commented:
@thekad I think I covered everything.

shaior commented on roles/junit2json/tasks/merge.yml:
```diff
@@ -0,0 +1,54 @@
+  changed_when:
+    - true
+
+- name: Validate some variables
```
Hi Max, 
Few comments here please - 

1. The task name does not specify the real intention of this task. 
2. I would put this assertion somewhere at the beginning of the playbook, so that if this package is missing it will fail on an early stage (as @thekad) wrote. 
3. To me it looks redundant to register a variable to represent package installation, there are more efficient, readable methods to do this, WDYT?

shaior commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,74 @@
+        path: "{{ junit2json_output_dir }}"
+        state: directory
+        mode: '0750'
+    - name: Write files (data + hash file)
```
Please insert an empty line between the two tasks for better readability.

thekad commented on roles/junit2json/defaults/main.yml:
```diff
@@ -0,0 +1,13 @@
+global_json_reports_list: []
+
+# defaults:
+junit2json_input_merged_report: ''
```
already left a note up in README, according to the docs this is only required if `do_merge` is true (the default in line 13) so either you make this a sane default, or remove the default and make it required because right now it is bad UX

thekad commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,78 @@
+  ansible.builtin.set_fact:
+    junit2json_result_data: "{{ junit2json_xml_report_content | redhatci.ocp.junit2obj | to_nice_json }}"
+  when:
+    - not(junit2json_do_merge)
```
parenthesis aren't needed here

thekad commented on roles/junit2json/tasks/merge.yml:
```diff
@@ -0,0 +1,40 @@
+    msg:
+      - "XML files names: {{ junit2json_reports_list }}"
+  when:
+    - junit2json_debug is defined
```
`junit2json_debug` is part of your defaults, it will always be defined, no need for this first check

thekad commented on roles/junit2json/vars/main.yml:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
```
if the entire file is commented out just remove it completely

mvk commented:
retest

ramperher commented on roles/junit2json/examples/example_playbook.yml:
```diff
@@ -0,0 +1,19 @@
+        junit2json_input_reports_list: "{{ input_reports | list }}"
+        junit2json_output_dir: "{{ playbook_dir }}/json_reports"
+        junit2json_do_merge: false
+    - name: Print the resulting list of JSON files
```
Break line before this task for better readability

ramperher commented on roles/junit2json/meta/argument_specs.yml:
```diff
@@ -0,0 +1,47 @@
+      New filename(s) is(are) based on the old ones and stored in global variable `global_json_reports_list`.
+    author:
+      - Max Kovgan
+    options:
```
What about `global_json_reports_list`? This is in defaults file. Should it appear here?

ramperher commented on roles/junit2json/tasks/main.yml:
```diff
@@ -0,0 +1,61 @@
+    - package: junitparser
+      exec: junitparser
+
+- name: Check whether role python dependency is installed
```
Isn't this task and the next one already covered with validate-dependency.yml?

ramperher commented on roles/junit2json/tasks/validate.dependency.yml:
I think it's better to call this play validate-dependency.yml instead

ramperher commented on roles/junit2json/tasks/main.yml:
```diff
@@ -0,0 +1,61 @@
+# NOTE: in case somebody did not install the collection properly, we test for filter's dependency
+
+- name: Check dependencies
+  ansible.builtin.include_tasks: validate.dependency.yml
```
Move to validate-dependency.yml file name

thekad commented:
all of my feedback got addressed, I'll let @ramperher approve

ramperher commented on meta/execution-environment.yml:
@mvk , this has not been addressed, could you please clarify?

ramperher reviewed (dismissed):
LGTM

ramperher commented:
> all of my feedback got addressed, I'll let @ramperher approve

Fine from my side, I think it needs a review from your side before merging since it's requiring @redhatci/maintain

tonyskapunk commented on README.md:
```diff
@@ -147,6 +148,7 @@ Name | Description
 Name | Type | Description
 --- | --- | ---
 [redhatci.ocp.junit2dict]() | Filter | Transforms a JUnit into a dictionary
+[redhatci.ocp.junit2obj]() | Filter | Transforms a JUnit into a corresponding JSON text (will replace `redhatci.ocp.junit2dict`)
```
Could you please remove this entry with references with `redhatci.ocp.junit2dict` in https://github.com/redhatci/ansible-collection-redhatci-ocp/pull/510#issuecomment-2553454615 was mentioned that the expectation is to remove the plugin rather than replace it by junit2obj.

tonyskapunk reviewed (commented):
Have you tested the role and/or the example provided? Do you mind sharing the results?

mvk commented:
> Have you tested the role and/or the example provided? Do you mind sharing the results?

1. clone the repo locally under `~/src/github.com/redhatci/collections/ansible_collections/redhatci/ocp` folder
1. run: `cd ~/src/github.com/redhatci/collections/ansible_collections/redhatci/ocp`
1. run:
```shell
cat > ansible.cfg << _EOF
[defaults]
collections_path = ../../../../collections
_EOF
```
4. create virtualenv, install the packages from `meta/requirements.txt` 
5. run inside the virtualenv:
```bash
ansible-playbook -i roles/junit2json/tests/inventory roles/junit2json/tests/test.yml
```

mvk commented:
[junit2json-tests.log](https://github.com/user-attachments/files/19315569/junit2json-tests.log)

tonyskapunk reviewed (changes requested):
Thanks for running the test and making the necessary changes to make it work.

Did you try the merge functionality? It does not seem to work.

I'm providing some changes to make the produced files return JSON and not a string of json.

mvk commented:
> Thanks for running the test and making the necessary changes to make it work.
> 
> Did you try the merge functionality? It does not seem to work.
> 
> I'm providing some changes to make the produced files return JSON and not a string of json.

@tonyskapunk thank you!!! 
(not enough tests after refactoring).

thanks to you, discoveries:
### the bug you found
added test case and fixed it as it was broken.
### `junitparser` needs to go asap
I used these files as input reports:
- `tests/unit/data/test_junit2obj_simple_input.xml`
- `tests/unit/data/test_junit2obj_failure_input.xml

to construct total run time I ran this in ipython:
```
In [1]: durations = [[0.026203, 0.017825], [0.0, 0.0, 6.826e-06]]

In [2]: sum([sum(i) for i in durations])
Out[5]: 0.044034826
```
But the test failed.
I looked at the file that `junitparser` generated, and it had the duration of `0.044`
👍

tonyskapunk commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,78 @@
+
+- name: Update junit2json_result_data junit2json_do_merge=true
+  ansible.builtin.set_fact:
+    junit2json_result_data: "{{ _junit2json_merged_xml.stdout | redhatci.ocp.junit2obj | to_nice_json }}"
```
The produced files are still a string with json, this suggestion remains valid.

tonyskapunk replied:
@mvk PTAL, the produced files are still a string with json.

tonyskapunk commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,78 @@
+
+- name: Convert junit XML to JSON and save in junit2json_result_data
+  ansible.builtin.set_fact:
+    junit2json_result_data: "{{ junit2json_xml_report_content | redhatci.ocp.junit2obj | to_nice_json }}"
```
ditto

tonyskapunk commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,78 @@
+        dest: "{{ item.dest }}"
+        mode: "{{ item.mode }}"
+      loop:
+        - content: "{{ junit2json_result_data | ansible.builtin.from_json | to_nice_json }}"
```
ditto

tonyskapunk replied:
@mvk

mvk commented:
> @mvk please address the questions and feedback provided to continue moving forward with this change, thanks

I do not see any unanswered questions/feedback.
Am I missing something you see and I do not ?

tonyskapunk commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,78 @@
+- name: Generate content hash from the content
+  ansible.builtin.set_fact:
+    junit2json_xml_report_hash_path: "{{ item }}.{{ junit2json_hash }}"
+    junit2json_xml_report_hash_curr: "{{ junit2json_xml_report_content | ansible.builtin.hash(junit2json_hash) }}"
```
@mvk can you provide a respond to this question?

tonyskapunk commented:
> I do not see any unanswered questions/feedback. Am I missing something you see and I do not ?

I've tagged you on them

mvk commented:
@tonyskapunk , please re-review, all your comments/questions are addressed.

mvk commented:
> > I do not see any unanswered questions/feedback. Am I missing something you see and I do not ?
> 
> I've tagged you on them

updated the pr, addressed all the comments.

ramperher commented on roles/junit2json/tasks/merge.yml:
```diff
@@ -0,0 +1,27 @@
+    dest: "{{ junit2json_output_dir }}/{{ junit2json_input_merged_report }}" # so that our findings don't find it
+    mode: '0644'
+  when:
+    - _junit2json_merged_xml.stdout | length > 10
```
Why this 10 here? What's the meaning?
If that's a constant value, maybe this should be defined somewhere else and use it accordingly.

ramperher reviewed (commented):
@mvk , overall the change looks fine for me, I've left one more comment for something that I think needs to be reviewed.
Before merging, note you need to squash your commits so that you only have one single commit in the PR.

tonyskapunk commented on plugins/filter/junit2obj.py:
```diff
@@ -30,9 +30,10 @@
-    - extra_metadata:
-      description: additional metadata dict added at the root of resulting JSON object
-      type: dict
+    - object:
```
Options[1] expect key values, not a list, `xml_report_text` also needs to be changed.
```suggestion
    object:
```

This is required or the documentation won't be rendered.

[1] https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#documentation-fields

tonyskapunk commented on plugins/filter/junit2obj.py:
```diff
@@ -19,9 +19,9 @@
 DOCUMENTATION = r"""
   name: junit2obj
   version_added: "1.0.0"
```
With the changes provided this would be a new version, right?

tonyskapunk commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,36 @@
+
+- name: Update junit2json_result_data junit2json_do_merge=true
+  ansible.builtin.set_fact:
+    junit2json_result_data: "{{ _junit2json_merged_xml.stdout if junit2json_do_merge else junit2json_xml_report_content }}"
```
nit, in cases like this use ternary:
```suggestion
    junit2json_result_data: "{{ junit2json_do_merge | ternary(_junit2json_merged_xml.stdout, junit2json_xml_report_content) }}"
```

tonyskapunk commented on roles/junit2json/tasks/convert.yml:
```diff
@@ -0,0 +1,36 @@
+
+- name: Set junit2json_output_report_path
+  ansible.builtin.set_fact:
+    junit2json_output_report_path: "{{ junit2json_output_dir }}/{{ junit2json_output_merged_report if junit2json_do_merge else junit2json_json_report_path }}"
```
nit:
```suggestion
   junit2json_output_report_path: "{{ junit2json_output_dir }}/{{ junit2json_do_merge | ternary(junit2json_output_merged_report,junit2json_json_report_path) }}"
```

tonyskapunk commented on roles/junit2json/tasks/expand.yml:
```diff
@@ -0,0 +1,27 @@
+
+- name: Print file_name value
+  ansible.builtin.debug:
+    msg: "file_name: {{ file_name }}"
```
nit:
```suggestion
    var: file_name
```

tonyskapunk commented on roles/junit2json/tasks/main.yml:
```diff
@@ -0,0 +1,47 @@
+
+- name: Print input reports variable
+  ansible.builtin.debug:
+    msg: "junit2json_input_reports_list: {{ junit2json_input_reports_list }}"
```
nit:
```suggestion
    var: junit2json_input_reports_list

tonyskapunk reviewed (changes requested):
There's only one required change, the other ones are nitpicks, up to you if you want to change them.

tonyskapunk commented on roles/junit2json/tasks/merge.yml:
```diff
@@ -0,0 +1,31 @@
+    dest: "{{ junit2_output_dir }}/{{ junit2_input_merged_report }}" # so that our findings don't find it
+    mode: '0644'
+  when:
+    - _junit2json_merged_xml.stdout | length > junit2_xml_min_length
```
```suggestion
    - _junit2json_merged_xml.stdout | length > junit2_xml_min_length | int
```

ramperher commented:
@mvk , changes looks good, you only need to squash your commits to only have one single commit in the PR. Thanks.

mvk commented:
@ramperher I squashed. is it a documented requirement? the 3 commits were different in the location.

ramperher commented:
> @ramperher I squashed. is it a documented requirement? the 3 commits were different in the location.

Yes, it's a requirement, we have this documented in https://github.com/redhatci/ansible-collection-redhatci-ocp/blob/main/CONTRIBUTING.md#pull-requests: _Squash the commits when you are ready to merge the PR._

I will accept it now, thanks!

ramperher reviewed (approved):
LGTM

tonyskapunk reviewed (approved):
LGTM, thanks for the effort.

mvk commented:
retest

tonyskapunk commented:
retest
